### PR TITLE
Support user overrides of three more framework-provided `ModelState`-related messages

### PR DESCRIPTION
**Pull request by dougbu**
- #3215
- add new accessor properties to `IModelBindingMessageProvider` and plumb them through
  - use in `ModelStateDictionary` when handling a `FormatException` or `OverflowException`
  - use in `ValidationHelpers` when handling a `ModelError` with `null` `ErrorMessage`
- add new `ModelExplorer` parameter to `IHtmlGenerator.GenerateValidationMessage()`
  - plumb through to `ValidationHelpers.GetModelErrorMessageOrDefault()`

Started from work @kichalla did on the `kiran/movemessages-to-messageprovider` branch in #3775.

nits:
- use helper methods more consistently in `HtmlHelper<T>`; slightly improves error checking
- remove unused `Resources` class from `Microsoft.AspNet.Mvc`
- make `ValidationHelpers` class `public`; already in `.Internal` namespace
  - split `GetUserErrorMessageOrDefault()` in two; rename to `GetModelErrorMessageOrDefault()`
- fix some #YOLO wrapping

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Properties/Resources.Designer.cs [hunk 1]
```diff
@@ -475,7 +475,7 @@ internal static string FormatAttributeRoute_TokenReplacement_UnescapedBraceInTok
         }
 
         /// <summary>
-        /// Unable to find the required services. Please add all the required services by calling '{0}.{1}' inside the call to '{4}' in the application startup code.
+        /// Unable to find the required services. Please add all the required services by calling '{0}.{1}' inside the call to '{2}' in the application startup code.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Properties/Resources.Designer.cs]: GitHub hides changes here by default. But it looks like the `resx` target hasn't run recently.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs [hunk 2]
```diff
@@ -3,27 +3,45 @@
...
 using System.Linq;
 using Microsoft.AspNet.Mvc.ModelBinding;
 
 namespace Microsoft.AspNet.Mvc.ViewFeatures.Internal
 {
-    internal static class ValidationHelpers
+    public static class ValidationHelpers
     {
-        public static string GetUserErrorMessageOrDefault(ModelError modelError, ModelStateEntry entry)
+        public static string GetModelErrorMessageOrDefault(ModelError modelError)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: Not sure what a "user" error meant.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 3]
```diff
@@ -96,10 +99,13 @@ public override void Contextualize(ViewContext viewContext)
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = ExpressionMetadataProvider.FromLambdaExpression(expression, ViewData, MetadataProvider);
-
-            return GenerateDropDown(modelExplorer, ExpressionHelper.GetExpressionText(expression), selectList,
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: Methods in this class were fairly random about `ExpressionMetadataProvider.FromLambdaExpression()` vs `GetModelExplorer()` and `ExpressionHelper.GetExpressionText()` vs `GetExpressionName()`. Change to use the helpers almost everywhere did not affect any tests though the helpers have a couple of additional error cases.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs [hunk 4]
```diff
@@ -1135,17 +1140,19 @@ protected virtual string GenerateName(string expression)
         }
 
         protected virtual IHtmlContent GenerateValidationMessage(
+            ModelExplorer modelExplorer,
             string expression,
             string message,
-            object htmlAttributes,
-            string tag)
+            string tag,
+            object htmlAttributes)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs]: Switched the order of these parameters to better match `IHtmlGenerator.GenerateValidationMessage()`.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs [hunk 5]
```diff
@@ -69,8 +69,8 @@ public interface IHtmlGenerator
         /// Generate a &lt;input type="checkbox".../&gt; element.
         /// </summary>
         /// <param name="viewContext">The <see cref="ViewContext"/> instance for the current scope.</param>
-        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> for the model.</param>
-        /// <param name="expression">The model expression.</param>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> for the <paramref name="expression"/>.</param>
+        /// <param name="expression">Expression name, relative to the current model.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: `GenerateCheckbox()` was not like the others ... for no good reason.

File: src/Microsoft.AspNet.Mvc/Properties/Resources.Designer.cs [hunk 6]
```diff
@@ -1,45 +0,0 @@
-// <auto-generated />
-namespace Microsoft.AspNet.Mvc
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/Properties/Resources.Designer.cs]: Checked-in generated file wasn't deleted when the `Resources.resx` file was.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: User == developer, I think. Fine either way.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs [hunk 7]
```diff
@@ -3,27 +3,45 @@
...
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using Microsoft.AspNet.Mvc.ModelBinding;
 
 namespace Microsoft.AspNet.Mvc.ViewFeatures.Internal
 {
-    internal static class ValidationHelpers
+    public static class ValidationHelpers
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: Ugh.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs [hunk 8]
```diff
@@ -3,27 +3,45 @@
...
+
+        public static string GetModelErrorMessageOrDefault(
+            ModelError modelError,
+            ModelStateEntry containingEntry,
+            ModelExplorer modelExplorer)
+        {
+            Debug.Assert(modelError != null);
+            Debug.Assert(containingEntry != null);
+            Debug.Assert(modelExplorer != null);
+            Debug.Assert(modelExplorer.Metadata?.ModelBindingMessageProvider?.ValueIsInvalidAccessor != null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: nit: odd to put asserts about an unrelated class here. We're not trying to make the code provably correct :+1:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs]: Was this method here in MVC5?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: I'm trusting you that you're doing the right thing here. I feel like you're the only person that really knows what that is...

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: You want the class to be renamed as well? Or are you objecting to something else?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Internal/ValidationHelpers.cs]: :ok: I'll remove the last `Assert()`.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs]: No

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: thanx

**Comment by dougbu** [comment]
04453a2

**Pull request closed without merging**